Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('heart_pandas.csv')

In [3]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [9]:
## Ваш код
sex_count = data.sex.value_counts()

In [13]:
## Ваш ответ:
print('Количество мужчин - %d' % (sex_count.male),
      'Количество женщин - %d' % (sex_count.female), sep ='\n')

Количество мужчин - 207
Количество женщин - 96


 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [47]:
## Ваш код
percent_sex = data.sex.value_counts(normalize = True).mul(100)

In [39]:
## Ваш ответ:
print ('Процент мужчин - %.2f' % (percent_sex.male), '%')

Процент мужчин - 68.32 %


**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [49]:
## Ваш код
heart_dis = data.groupby('sex').target.sum()

In [51]:
## Ваш ответ:
print('Мужчины с заболеванием сердца - %d человека' % (heart_dis.male),
      'Женщины с заболеванием сердца  - %d человека' % (heart_dis.female), sep ='\n')

Мужчины с заболеванием сердца - 93 человека
Женщины с заболеванием сердца  - 72 человека


**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [53]:
## Ваш код
male0 = data[data.sex == 'male'].target[data.target == 0].count()
m = male0/len(data)

In [54]:
## Ваш ответ:
print ('Доля мужчин, не имеющих болезнь сердца - %.2f' % (m))

Доля мужчин, не имеющих болезнь сердца - 0.38


**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [55]:
## Ваш код
young = data[data.target == 1].age.min()

In [57]:
## Ваш ответ:
print ('Возраст самого молодого пациента с болезнью сердца - %.d' % (young))

Возраст самого молодого пациента с болезнь сердца - 29


**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [59]:
## Ваш код
old = data[data.target == 0].age.max()

In [60]:
## Ваш ответ:
print ('Возраст самого старого пациента без болезни сердца - %.d' % (old))

Возраст самого старого пациента без болезни сердца - 77


**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [61]:
## Ваш код
young_woman = data.query('target == 1 and sex == "female"').age.min()

In [62]:
## Ваш ответ:
print ('Возраст самой молодой женщины с болезнью сердца - %.d' % (young_woman))

Возраст самой молодой женщины с болезнью сердца - 34


**Какой средний возраст женщин?**

In [64]:
## Ваш код
fem_age = data[data.sex == 'female'].age.mean()

In [122]:
## Ваш ответ:
print ('Средний возраст женщин - %.2f' % (fem_age))

Средний возраст женщин - 55.68


**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [67]:
## Ваш код
av = data[data.target == 1].age.mean()
st_dev = data[data.target == 1].age.std()

In [71]:
## Ваш ответ:
print ('Средний возраст - %.2f' % (av),
       'Среднеквадратическое отклонение возраста %.2f' % (st_dev), sep ='\n')


Средний возраст - 52.50
Среднеквадратическое отклонение возраста 9.55


** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [73]:
## Ваш код
average_chol = data.cholesterol.mean()
average0 = data[data.target == 0].cholesterol.mean()
if average0 < average_chol:
    stment = True
else:
    stment = False

In [74]:
## Ваш ответ:
print ('Данное утверждение -', stment)

Данное утверждение - False


**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [83]:
## Ваш код
data.groupby('rest_ecg')['age',
                         'resting_blood_pressure',
                         'cholesterol', 
                         'max_heart_rate_achieved',
                         'st_depression'].agg(['mean','max'])


C:\Users\Александра\AppData\Local\Temp\ipykernel_335976\2528024134.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  data.groupby('rest_ecg')['age','resting_blood_pressure','cholesterol', 'max_heart_rate_achieved','st_depression'].agg(['mean','max'])


age     resting_blood_pressure       \
                                   mean max                   mean  max   
rest_ecg                                                                  
ST-T wave abnormality         52.914474  71             129.065789  180   
left ventricular hypertrophy  61.000000  76             140.500000  180   
normal                        55.687075  77             134.027211  200   

                             cholesterol      max_heart_rate_achieved       \
                                    mean  max                    mean  max   
rest_ecg                                                                     
ST-T wave abnormality         237.269737  354              151.960526  194   
left ventricular hypertrophy  261.750000  327              125.750000  140   
normal                        255.142857  564              147.904762  202   

                             st_depression       
                                      mean  max  
rest_ecg                                         
ST-T wave abnormality             0.879605  5.6  
left ventricular hypertrophy      2.725000  4.4  
normal                            1.159184  6.2

In [35]:
## Ваш ответ:

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [92]:
## Ваш код
data.groupby(['sex','target']).st_depression.mean()

sex     target
female  0         1.841667
        1         0.554167
male    0         1.531579
        1         0.605376
Name: st_depression, dtype: float64

In [121]:
## Ваш ответ:
fem0 = data.query('target == 0 and sex == "female"').st_depression.mean()
mal1 = data.query('target == 1 and sex == "male"').st_depression.mean()
if mal1>fem0:
    print('Уровень депрессии у мужчин с болезнью сердца (%.2f) выше, чем у женщин, не страдающих болезнью сердца (%.2f)'
         % (mal1, fem0))
else:
    print('Уровень депрессии у мужчин с болезнью сердца (%.2f) ниже, чем у женщин, не страдающих болезнью сердца (%.2f)'
         % (mal1, fem0))

Уровень депрессии у мужчин с болезнью сердца (0.61) ниже, чем у женщин, не страдающих болезнью сердца (1.84)


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [79]:
## Ваш код
tp = ['chest_pain_type','rest_ecg','thalassemia']
chl = pd.DataFrame(columns = ['max', 'min'])
for i in tp:
    chl_i = data.groupby(i).cholesterol.agg(['max','min'])
    chl = pd.concat((chl, chl_i))

In [80]:
## Ваш ответ:
print(chl)

                              max  min
atypical angina               564  126
non-anginal pain              298  182
typical angina                409  131
ST-T wave abnormality         354  126
left ventricular hypertrophy  327  197
normal                        564  149
fixed defect                  417  141
normal                        318  169
reversable defect             564  126


**Сколько значений может принимать каждый из категориальных признаков?**

In [81]:
## Ваш код
catg = ['sex', 
        'chest_pain_type', 
        'rest_ecg', 
        'exercise_induced_angina', 
        'thalassemia', 
        'target']
un_catg = data[catg].nunique()

In [82]:
## Ваш ответ:
print(un_catg)

sex                        2
chest_pain_type            3
rest_ecg                   3
exercise_induced_angina    2
thalassemia                3
target                     2
dtype: int64


**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [111]:
## Ваш код
catg = ['sex', 
        'chest_pain_type', 
        'rest_ecg', 
        'exercise_induced_angina', 
        'thalassemia', 
        'target']
disb = 0
for i in catg:
    count_catg = data[i].value_counts()
    diff = count_catg.max() - count_catg.min()
    if disb < diff:
        disb = diff
        disb_catg = i

In [110]:
## Ваш ответ:
print ('Cамый сильный дисбаланс классов у признака:', disb_catg)

Cамый сильный дисбаланс классов у признака: chest_pain_type


---